### PR TITLE
CY-644 Capabilities (WIP)

### PR DESCRIPTION
**Pull request by mcouthon**

### CODE REVIEW

File: cloudify_rest_client/deployments.py [hunk 1]
```diff
@@ -124,6 +124,23 @@ def outputs(self):
...
 
+class DeploymentCapabilities(dict):
+
+    def __init__(self, outputs):
+        super(DeploymentCapabilities, self).__init__()
+        self.update(outputs)
+
+    @property
+    def deployment_id(self):
+        """Deployment Id the outputs belong to."""
```
Comment (tehasdf) [inline cloudify_rest_client/deployments.py]: s/outputs/capabilities/

File: cloudify/workflows/local.py [hunk 2]
```diff
@@ -89,7 +89,8 @@ def outputs(self):
             get_node_instances_method=self.storage.get_node_instances,
             get_node_instance_method=self.storage.get_node_instance,
             get_node_method=self.storage.get_node,
-            get_secret_method=self.storage.get_secret)
+            get_secret_method=self.storage.get_secret,
+            get_capability_method=self.storage.get_capability)
```
Comment (tehasdf) [inline cloudify/workflows/local.py]: not implemented yet, right?

File: dsl_parser/functions.py [hunk 3]
```diff
@@ -489,6 +499,48 @@ def evaluate_runtime(self, storage):
...
+    def __init__(self, args, **kwargs):
+        self.capability_path = None
+        super(GetCapability, self).__init__(args, **kwargs)
+
+    def parse_args(self, args):
+        if not isinstance(args, list):
+            raise ValueError(
+                "`get_capability` function argument should be a list. Instead "
+                "it is a {0} with the value: {1}.".format(type(args), args))
+        if not len(args) == 2:
```
Comment (tehasdf) [inline dsl_parser/functions.py]: s/`not ==`/`!=`/

File: dsl_parser/functions.py [hunk 4]
```diff
@@ -634,18 +689,47 @@ def evaluate_functions(payload, context,
...
+
+    :param capabilities: The dict of capabilities to evaluate
+    :param get_node_instances_method: A method for getting node instances.
+    :param get_node_instance_method: A method for getting a node instance.
+    :param get_node_method: A method for getting a node.
+    :param get_secret_method: A method for getting a secret.
+    :param get_capability_method: A method for getting a capability.
+    :return: Outputs dict.
+    """
+    capabilities = dict((k, v['value']) for k, v in capabilities.iteritems())
```
Comment (tehasdf) [inline dsl_parser/functions.py]: i'd say let's use `.items()` in new code so that we don't add more things to change when porting to py3

Comment (tehasdf) [reply dsl_parser/functions.py]: also, this always runs on 2.7, right? ie. not on the agent
use a dictcomp if so

File: dsl_parser/tests/test_get_capability.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
+########
+# Copyright (c) 2018 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (tehasdf) [inline dsl_parser/tests/test_get_capability.py]: cloudify platform ltd

File: cloudify/workflows/local.py (hunk 2 above)
Comment (mcouthon) [reply cloudify/workflows/local.py]: It is, in https://github.com/cloudify-cosmo/cloudify-manager/pull/1490

File: dsl_parser/functions.py [hunk 6]
```diff
@@ -634,18 +689,47 @@ def evaluate_functions(payload, context,
...
+                          get_capability_method):
+    """Evaluates capabilities definition containing intrinsic functions.
+
+    :param capabilities: The dict of capabilities to evaluate
+    :param get_node_instances_method: A method for getting node instances.
+    :param get_node_instance_method: A method for getting a node instance.
+    :param get_node_method: A method for getting a node.
+    :param get_secret_method: A method for getting a secret.
+    :param get_capability_method: A method for getting a capability.
+    :return: Outputs dict.
```
Comment (tyacbovi) [inline dsl_parser/functions.py]: Outputs -> capabilities

File: dsl_parser/tests/test_get_capability.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        except exception_type as e:
+            self.assertIn(message, str(e))
+        else:
+            raise AssertionError('Error was not raised')
+
+    def _assert_parsing_fails(self, yaml, message):
+        self._assert_raises_with_message(
+            ValueError,
+            message,
+            self.parse_1_3,
```
Comment (tyacbovi) [inline dsl_parser/tests/test_get_capability.py]: why not use self.parse?

File: dsl_parser/tests/test_get_capability.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+from dsl_parser import functions
+from dsl_parser.tasks import prepare_deployment_plan
+from dsl_parser.tests.abstract_test_parser import AbstractTestParser
+
+
+class TestGetCapability(AbstractTestParser):
```
Comment (tyacbovi) [inline dsl_parser/tests/test_get_capability.py]: Is it possible to set a default value for an input with a capability?

Comment (tyacbovi) [inline dsl_parser/tests/test_get_capability.py]: Maybe add tests on invalid names for capability and tests on a not existing capability

File: dsl_parser/tests/test_get_capability.py (hunk 7 above)
Comment (mcouthon) [reply dsl_parser/tests/test_get_capability.py]: Because the default there is v1.0.

File: dsl_parser/tests/test_get_capability.py (hunk 8 above)
Comment (mcouthon) [reply dsl_parser/tests/test_get_capability.py]: What's the difference between invalid name and a non existing capability?

Comment (mcouthon) [reply dsl_parser/tests/test_get_capability.py]: Anyway, I'll validate this in the manager's unittests, as this more closely resembles a real env.

Comment (mcouthon) [reply dsl_parser/tests/test_get_capability.py]: Apparently yes! I'll add a test to validate.

**Pull request merged**